File: packages/javascript/src/internal/drawing.js
const MAX_SURFACE_AXIS = 2_048;
const MAX_RGB24 = 0xff_ff_ff;
const MIN_CONVERTED_WIDTH = 1 / 256;
const hasOwn = (value, key) => Object.prototype.hasOwnProperty.call(value, key);
const binary64View = new DataView(new ArrayBuffer(8));

/** Error with a stable fresh-raster validation code. */
export class DrawingError extends Error {
  constructor(code) {
    super(code);
    this.name = "DrawingError";
    this.code = code;
  }
}

function fail(code) {
  throw new DrawingError(code);
}

function normalizedZero(value) {
  return value === 0 ? 0 : value;
}

function dataProperty(object, key) {
  const descriptor = Object.getOwnPropertyDescriptor(object, key);
  return descriptor !== undefined && "value" in descriptor ? descriptor.value : undefined;
}

function hasOnlyOwnDataKeys(value, expected) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) return false;
  const keys = Reflect.ownKeys(value);
  return keys.length === expected.length && expected.every((key) => hasOwn(value, key));
}

function finiteNumber(value) {
  return typeof value === "number" && Number.isFinite(value);
}

function boundedInteger(value, minimum, maximum) {
  return finiteNumber(value) && Number.isInteger(value) && value >= minimum && value <= maximum;
}

function exactArray(value, length) {
  if (!Array.isArray(value) || value.length !== length) return false;
  const keys = Reflect.ownKeys(value);
  if (keys.length !== length + 1 || !keys.includes("length")) return false;
  for (let index = 0; index < length; index += 1) {
    if (!hasOwn(value, index) || !finiteNumber(dataProperty(value, String(index)))) return false;
  }
  return true;
}

function normalizedPoint(value) {
  if (!exactArray(value, 2)) fail("INVALID_COMMAND");
  return [normalizedZero(dataProperty(value, "0")), normalizedZero(dataProperty(value, "1"))];
}

// An exact represented-binary64 value is significand * 2 ** exponent. Keeping
// the exponent separate makes orientation signs exact without decimal conversion
// or an epsilon, including subnormals and large cancellation-prone coordinates.
function binary64Rational(value) {
  binary64View.setFloat64(0, value, false);
  const bits = binary64View.getBigUint64(0, false);
  const sign = (bits >> 63n) === 0n ? 1n : -1n;
  const exponentBits = Number((bits >> 52n) & 0x7ffn);
  const fraction = bits & ((1n << 52n) - 1n);
  if (exponentBits === 0) return { significand: sign * fraction, exponent: -1074 };
  return { significand: sign * ((1n << 52n) | fraction), exponent: exponentBits - 1075 };
}

function subtractExact(left, right) {
  const exponent = Math.min(left.exponent, right.exponent);
  return {
    significand: (left.significand << BigInt(left.exponent - exponent)) -
      (right.significand << BigInt(right.exponent - exponent)),
    exponent,
  };
}

function multiplyExact(left, right) {
  return { significand: left.significand * right.significand, exponent: left.exponent + right.exponent };
}

function orientSign(a, b, c) {
  const ax = binary64Rational(a[0]);
  const ay = binary64Rational(a[1]);
  const bx = binary64Rational(b[0]);
  const by = binary64Rational(b[1]);
  const cx = binary64Rational(c[0]);
  const cy = binary64Rational(c[1]);
  const first = multiplyExact(subtractExact(bx, ax), subtractExact(cy, ay));
  const second = multiplyExact(subtractExact(by, ay), subtractExact(cx, ax));
  const result = subtractExact(first, second).significand;
  return result > 0n ? 1 : result < 0n ? -1 : 0;
}

/**
 * Return whether four cyclic binary64 points have one nonzero exact turn sign.
 * Internal-only: it is exported from this module for native numeric conformance,
 * never from the package entry point.
 */
export function strictlyConvex(points) {
  let sign = 0;
  for (let index = 0; index < 4; index += 1) {
    const turn = orientSign(points[index], points[(index + 1) % 4], points[(index + 2) % 4]);
    if (turn === 0 || (sign !== 0 && turn !== sign)) return false;
    sign = turn;
  }
  return true;
}

function binary32(value) {
  const converted = Math.fround(value);
  if (!Number.isFinite(converted)) fail("INVALID_COMMAND");
  return normalizedZero(converted);
}

function convertedPoint(point) {
  return [binary32(point[0]), binary32(point[1])];
}

/** Validate and detach the fresh density-1 drawing environment. */
export function validateEnvironment(value) {
  const expected = ["width", "height", "density", "background"];
  if (!hasOnlyOwnDataKeys(value, expected)) fail("INVALID_ENVIRONMENT");
  const width = dataProperty(value, "width");
  const height = dataProperty(value, "height");
  const density = dataProperty(value, "density");
  const background = dataProperty(value, "background");
  if (!boundedInteger(width, 1, MAX_SURFACE_AXIS) || !boundedInteger(height, 1, MAX_SURFACE_AXIS) ||
      density !== 1 || !boundedInteger(background, 0, MAX_RGB24)) {
    fail("INVALID_ENVIRONMENT");
  }
  return { width, height, density: 1, background: normalizedZero(background) };
}

function validateStyle(command) {
  const rgb = dataProperty(command, "rgb");
  const opacity8 = dataProperty(command, "opacity8");
  if (!boundedInteger(rgb, 0, MAX_RGB24) || !boundedInteger(opacity8, 0, 255)) {
    fail("INVALID_COMMAND");
  }
  return { rgb: normalizedZero(rgb), opacity8: normalizedZero(opacity8) };
}

function validateSegment(command) {
  if (!hasOnlyOwnDataKeys(command, ["kind", "from", "to", "rgb", "opacity8", "width", "cap"]) ||
      dataProperty(command, "kind") !== "segment2" || dataProperty(command, "cap") !== "round") {
    fail("INVALID_COMMAND");
  }
  const from = normalizedPoint(dataProperty(command, "from"));
  const to = normalizedPoint(dataProperty(command, "to"));
  const width = dataProperty(command, "width");
  if (!finiteNumber(width) || width <= 0 || (from[0] === to[0] && from[1] === to[1])) {
    fail("INVALID_COMMAND");
  }
  return { kind: "segment2", points: [from, to], width: normalizedZero(width), cap: "round", ...validateStyle(command) };
}

function validateQuad(command) {
  if (!hasOnlyOwnDataKeys(command, ["kind", "vertices", "rgb", "opacity8"]) ||
      dataProperty(command, "kind") !== "quad2") {
    fail("INVALID_COMMAND");
  }
  const vertices = dataProperty(command, "vertices");
  if (!Array.isArray(vertices) || vertices.length !== 4) fail("INVALID_COMMAND");
  const keys = Reflect.ownKeys(vertices);
  if (keys.length !== 5 || !keys.includes("length")) fail("INVALID_COMMAND");
  const points = new Array(4);
  for (let index = 0; index < 4; index += 1) {
    if (!hasOwn(vertices, index)) fail("INVALID_COMMAND");
    points[index] = normalizedPoint(dataProperty(vertices, String(index)));
  }
  if (!strictlyConvex(points)) fail("INVALID_COMMAND");
  return { kind: "quad2", points, ...validateStyle(command) };
}

function inConvertedDomain(points, width, environment) {
  const margin = Math.max(environment.width, environment.height);
  for (const point of points) {
    if (point[0] < -margin || point[0] > environment.width + margin ||
        point[1] < -margin || point[1] > environment.height + margin) {
      return false;
    }
  }
  return width === undefined || (width >= MIN_CONVERTED_WIDTH && width <= margin);
}

/**
 * Normalize one command against an explicitly supplied fresh-raster environment.
 * The result is detached canonical data for a later renderer adapter; no surface,
 * state machine, host colour state, or raster operation is involved here.
 * The environment is revalidated here so this standalone internal helper remains
 * safe when called without a prior `validateEnvironment` step.
 */
export function normalizeCommand(command, environment) {
  const env = validateEnvironment(environment);
  if (!hasOnlyOwnDataKeys(command, ["kind", "from", "to", "rgb", "opacity8", "width", "cap"]) &&
      !hasOnlyOwnDataKeys(command, ["kind", "vertices", "rgb", "opacity8"])) {
    fail("INVALID_COMMAND");
  }

  const kind = dataProperty(command, "kind");
  const canonical = kind === "segment2" ? validateSegment(command) :
    kind === "quad2" ? validateQuad(command) : fail("INVALID_COMMAND");
  const points = canonical.points.map(convertedPoint);
  const width = canonical.kind === "segment2" ? binary32(canonical.width) : undefined;
  if (!inConvertedDomain(points, width, env) || (canonical.kind === "quad2" && !strictlyConvex(points))) {
    fail("INVALID_COMMAND");
  }

  const alpha64 = canonical.opacity8 / 255;
  const result = {
    outcome: canonical.kind === "segment2" && points[0][0] === points[1][0] && points[0][1] === points[1][1] ? "noop" : "emit",
    kind: canonical.kind,
    points,
    rgb: canonical.rgb,
    channels: [(canonical.rgb >>> 16) & 0xff, (canonical.rgb >>> 8) & 0xff, canonical.rgb & 0xff],
    opacity8: canonical.opacity8,
    alpha64: normalizedZero(alpha64),
  };
  if (canonical.kind === "segment2") {
    result.width = width;
    result.cap = "round";
  }
  return result;
}

File: packages/python/procedurals/_drawing.py
"""Internal pure validation for drawing.fresh-raster-2d v0.1.0.

Motivated by survey/out/2018/Generativos/pelines/notes.md and the ciserp
counterexample. Bounds are reviewed engineering policy, not artistic ranges.
No host objects or renderer support are provided by this module.
"""
from collections.abc import Mapping
import math
import struct


class DrawingError(ValueError):
    def __init__(self, code):
        self.code=code
        super().__init__(code)


def _number(value,code):
    if isinstance(value,bool) or not isinstance(value,(int,float)):
        raise DrawingError(code)
    try:value=float(value)
    except (OverflowError,ValueError):raise DrawingError(code) from None
    if not math.isfinite(value):raise DrawingError(code)
    return 0.0 if value==0 else value


def _integer(value,lo,hi,code):
    value=_number(value,code)
    if value!=math.floor(value) or not lo<=value<=hi:raise DrawingError(code)
    return int(value)


def validate_environment(value):
    """Return a detached environment; no allocation of a native surface."""
    code='INVALID_ENVIRONMENT'
    if not isinstance(value,Mapping) or set(value)!= {'width','height','density','background'}:
        raise DrawingError(code)
    return {'width':_integer(value['width'],1,2048,code),
            'height':_integer(value['height'],1,2048,code),
            'density':_integer(value['density'],1,1,code),
            'background':_integer(value['background'],0,16777215,code)}


def _point(value):
    if not isinstance(value,(list,tuple)) or len(value)!=2:raise DrawingError('INVALID_COMMAND')
    return [_number(value[0],'INVALID_COMMAND'),_number(value[1],'INVALID_COMMAND')]


def _convex(points):
    # Binary64 ratios have power-of-two denominators. Scale all coordinates to
    # one exact integer lattice; determinant signs are unchanged. No epsilon or
    # rounded multiply/subtract can erase a thin shape's nonzero turn.
    ratios=[v.as_integer_ratio() for p in points for v in p]
    denominator=max(d for _,d in ratios)
    coords=[n*(denominator//d) for n,d in ratios]
    sign=0
    for i in range(4):
        a=2*i;b=2*((i+1)%4);c=2*((i+2)%4)
        turn=(coords[b]-coords[a])*(coords[c+1]-coords[a+1])-(coords[b+1]-coords[a+1])*(coords[c]-coords[a])
        current=(turn>0)-(turn<0)
        if current==0 or (sign and current!=sign):return False
        sign=current
    return True


def _f32(value):
    try:result=struct.unpack('>f',struct.pack('>f',value))[0]
    except OverflowError:raise DrawingError('INVALID_COMMAND') from None
    if not math.isfinite(result):raise DrawingError('INVALID_COMMAND')
    return 0.0 if result==0 else result


def normalize_command(command,environment):
    """Normalize one command to detached values for a future adapter.

    The frame must already have validated its environment. This internal routine
    does not own a frame, advance indices, retain caller data or perform drawing.
    """
    code='INVALID_COMMAND'
    if not isinstance(command,Mapping):raise DrawingError(code)
    kind=command.get('kind')
    if kind=='segment2':
        if set(command)!={'kind','from','to','rgb','opacity8','width','cap'} or command['cap']!='round':
            raise DrawingError(code)
        points=[_point(command['from']),_point(command['to'])]
        width=_number(command['width'],code)
        if width<=0:raise DrawingError(code)
    elif kind=='quad2':
        if set(command)!={'kind','vertices','rgb','opacity8'}:raise DrawingError(code)
        vertices=command['vertices']
        if not isinstance(vertices,(list,tuple)) or len(vertices)!=4:raise DrawingError(code)
        points=[_point(p) for p in vertices];width=None
    else:raise DrawingError(code)
    rgb=_integer(command['rgb'],0,16777215,code)
    opacity=_integer(command['opacity8'],0,255,code)
    if (kind=='segment2' and points[0]==points[1]) or (kind=='quad2' and not _convex(points)):
        raise DrawingError(code)
    converted=[[_f32(x),_f32(y)] for x,y in points]
    if width is not None:width=_f32(width)
    w,h=environment['width'],environment['height'];m=max(w,h)
    if any(not (-m<=x<=w+m and -m<=y<=h+m) for x,y in converted):raise DrawingError(code)
    if width is not None and not 1/256<=width<=m:raise DrawingError(code)
    if kind=='quad2' and not _convex(converted):raise DrawingError(code)
    result={'outcome':'noop' if kind=='segment2' and converted[0]==converted[1] else 'emit',
            'kind':kind,'points':converted,'rgb':rgb,'channels':[(rgb>>16)&255,(rgb>>8)&255,rgb&255],
            'opacity8':opacity,'alpha64':opacity/255}
    if width is not None:result.update(width=width,cap='round')
    return result
